JOSS/paper.md

```diff
@@ -0,0 +1,103 @@
+---
+title: 'JLSO: Storing reproducible serialized objects in Julia'
+tags:
+  - julialang
+  - data storage
+  - reproducibility
+authors:
+  - name: Rory Finnegan
+    orcid: 0000-0002-4603-077X
+    affiliation: 1
+  - name: Lyndon White
+    orcid: 0000-0003-1386-1646
+    affiliation: 2
+affiliations:
+ - name: Research Engineer, Invenia Technical Computing
+   index: 1
+ - name: Research Engineer, Invenia Labs
+   index: 2
+date: 24 August 2022
+bibliography: paper.bib
+---
+
+# Summary
+
+As scientific computing software grows increasingly complex, the need to efficiently and reliably store sophisticated program objects has become a growing need. The expanding list of file formats for serializing objects is evidence of this problem. Unfortunately, these file formats typically come with usability versus reliability tradeoffs.
+
+Choosing serialization formats such as CSV, JSON, BSON, and HDF5 are prudent choices for long term storage because they are application and language agnostic. Projects that evolve beyond an initial language or library decision can still load old experimental data.
+These formats often support a limited set of types, requiring applications to define and maintain serialization methods for their custom objects.
+
+Using formats such as JLD (Julia Data), Python pickles or MAT files work best for convenient saving of arbitrary objects. Unfortunately, these formats are often highly coupled to the software dependencies (e.g., language version, software packages), which makes restoring the data more challenging as time passes and the software evolves.
```

Comment:
Probably we should mention specifically in this section that JLD and MAT are extensions of HDF5, and that while this has some advantages of HDF5, it falls back in other cases to being no better than pickle or the julia serializer.
We should mention fully bespoke things like Pickle and the julia serialiser first.

Reply:
We should make it clear that this JLSO is firmly inside the second category, and is vulnerable to all the same problems.
but that it seaks to amiliorate them via keeping enough metadata around to recover.

Reply:
Hmmm, should we maybe have 3 categories then?

1. General/agnostic file formats (e.g., CSV, BSON, HDF5)
2. Language specific serializers like pickles or the julia serializer
3. File formats that merge the two. Arguably JLSO fits into this category, but in a different way.

Didn't MAT files only extend the HDF spec in 7.3 though?

Reply:
Yes, but MatLab  7.3 came out 14 years ago now.

Reply:
I would say 1 and 2 exist.

and that .JLD, .MAT, and BSON.jl are special cases of 2
but that for the right inputs may be partially or totally readable with tooling for 1.

and JLSO is a related but different case to that.

Reply:
> MatLab 7.3 came out 14 years ago now.

Lol, good point.

> JLSO is a related but different case to that.

Okay, I've opted to remove the reference to JLD and MAT for now. I think this simplifies the discussion, but I can add it back if you think it's worth distinguishing JLSO from those specific formats.